pkg/util/hostname/providers.go

```diff
@@ -3,47 +3,193 @@
+		stopIfSucessful: true,
+		expvarName:      "'hostname_file' configuration/environment",
+	},
+	{
+		name:            "fargate",
+		cb:              fromFarget,
+		stopIfSucessful: true,
+		expvarName:      "fargate",
+	},
+	{
+		name:            "gce",
+		cb:              fromGCE,
+		stopIfSucessful: true,
+		expvarName:      "gce",
+	},
+	{
+		name:            "azure",
+		cb:              fromAzure,
+		stopIfSucessful: true,
+		expvarName:      "azure",
+	},
 
-// RegisterHostnameProvider registers a hostname provider as part of the catalog
-func RegisterHostnameProvider(name string, p Provider) {
-	providerCatalog[name] = p
+	// The following providers are coupled. Their behavior change following the result of the previous provider.
+	// Therefore 'stopIfSucessful' is set to false.
+	{
+		name:            "fqdn",
+		cb:              fromFQDN,
+		stopIfSucessful: false,
+		expvarName:      "fqdn",
+	},
+	{
+		name:            "container",
+		cb:              fromContainer,
+		stopIfSucessful: false,
+		expvarName:      "container",
+	},
+	{
+		name:            "os",
+		cb:              fromOS,
+		stopIfSucessful: false,
+		expvarName:      "os",
+	},
+	{
+		name:            "aws", // ie EC2
+		cb:              fromEC2,
+		stopIfSucessful: false,
+		expvarName:      "aws",
+	},
 }
 
-// GetProvider returns a Provider if it was register before.
-func GetProvider(providerName string) Provider {
-	if provider, found := providerCatalog[providerName]; found {
-		return provider
+// IsConfigurationProvider returns true if the hostname was found through the configuration file
+func (h Data) IsConfigurationProvider() bool {
+	return h.Hostname == configProvider
+}
+
+func saveHostname(cacheHostnameKey string, hostname string, providerName string) Data {
+	data := Data{
+		Hostname: hostname,
+		Provider: providerName,
+	}
+
+	cache.Cache.Set(cacheHostnameKey, data, cache.NoExpiration)
+	if providerName != "" && providerName != "fargate" {
+		hostnameProvider.Set(providerName)
+		inventories.SetAgentMetadata(inventories.AgentHostnameSource, providerName)
 	}
-	return nil
+	return data
 }
 
-// GetHostname returns the hostname for a specific Provider if it was register
-func GetHostname(ctx context.Context, providerName string, options map[string]interface{}) (string, error) {
-	if provider, found := providerCatalog[providerName]; found {
-		log.Debugf("GetHostname trying provider '%s' ...", providerName)
-		name, err := provider(ctx, options)
+// GetWithProvider returns the hostname for the Agent and the provider that was use to retrieve it
+func GetWithProvider(ctx context.Context) (Data, error) {
+	cacheHostnameKey := cache.BuildAgentKey("hostname")
+
+	// first check if we have a hostname cached
+	if cacheHostname, found := cache.Cache.Get(cacheHostnameKey); found {
+		return cacheHostname.(Data), nil
+	}
+
+	var err error
+	var hostname string
+	var providerName string
+	//providers := config.Datadog.GetStringSlice("hostname_providers")
```

Comment:
commented out?

Reply:
I first wanted to be able to configure which provider to use from the configuration. I quickly realized that would be for another PR since this one is already huge. I remove the comment